### PR TITLE
[WEB-457] Checkout Edit Address Updates

### PR DESCRIPTION
**Pull request by shaheen-karodia**
### What problem is this solving?
Currently for the Address Swarm Project - We need to update the following flows
- Add Address Manual
- Add Address Via Search
- Switch to Manual after Adding Address Via Search
- Edit Address

The problem with this approach is: Checkout uses all the same forms for all journeys.

This is reimplementation of the edit address flow in checkout in accordance with:
https://www.notion.so/bashdotcom/Address-Swarm-1-Checkout-Edit-Address-Flow-Updates-88978dc4627d401381449f7037d63f51?pvs=4

### How it works

**EditAddressForm**
-  We have split out 3 components for rendering:
 1. Heading
 2. SaveButton
 3. ContactCard 
 
 Edit Address Form
 -> Takes in an Address 
 -> Only Renders Selected Fields
 
 On Submit: 
 -> Full fields of the the form
 -> Validate the fields (apply UI to them if they are incorrect)
 -> Post The addresses to the UI, Local Storage, orderform, and Masterdata
 -> Switch to the Selected Addresses View

**Delivery Container**
A new section is added with `id="edit-adress-section" ` as a new mounting point for the Edit Form 

**Delivery Controller**
RenderEditAddress: -> takes in an address, and mounts a edit address form into the `edit-adress-section` mounting point

clearEditAddress -> clears the Edit Address from from the mounting point `edit-adress-section`

### How to test it?
https://addressswarmrefactor--thefoschini.myvtex.com/checkout/#/shipping

1. Go to check out from select address view click edit
image

2. You are presented with an updated edit address form 
image

3. On hitting save with invalid data - form validations witll kick in:
image

4 On Successful submission: the address overview will have selected your updated address and its update name/number will display
image

5. The orderforms selected address would have also applied
https://addressswarmrefactor--thefoschini.myvtex.com/api/checkout/pub/orderForm/5280d0421d9f49e89ce85c31821bc537

image
image

### Related to / Depends on
https://www.notion.so/bashdotcom/Address-Swarm-1-Checkout-Edit-Address-Flow-Updates-88978dc4627d401381449f7037d63f51?pvs=4

#### How does this PR make you feel? :link:
terrible - this sucks to write code in checkout

### CODE REVIEW

File: checkout-ui-custom/src/controller/DeliverController.js [hunk 1]
```diff
@@ -52,6 +53,17 @@ const DeliverController = (() => {
     }
   };
 
+  const RenderEditAddress = async (addressName) => {
+    const fields = 'id,receiverPhone,receiverName,number,companyBuilding,street,city,postalCode,state';
```
Comment (shaheen-karodia) [inline checkout-ui-custom/src/controller/DeliverController.js:57]: We need to look at getting addresses via Index DB and then only using the fields relevant to us

File: checkout-ui-custom/src/controller/DeliverController.js [hunk 2]
```diff
@@ -52,6 +53,17 @@ const DeliverController = (() => {
...
+    const data = await getAddress(addressName, `?_fields=${fields}`);
+    document.querySelector('#edit-adress-section').innerHTML = EditAddressForm(data);
+    preparePhoneField('#bash--input-receiverPhone');
```
Comment (shaheen-karodia) [inline checkout-ui-custom/src/controller/DeliverController.js:60]: We need to look at dropping the flag support

File: checkout-ui-custom/src/controller/DeliverController.js [hunk 3]
```diff
@@ -52,6 +53,17 @@ const DeliverController = (() => {
...
+  };
+
+  const clearEditAddress = () => {
```
Comment (shaheen-karodia) [inline checkout-ui-custom/src/controller/DeliverController.js:63]: Maybe one utility method for all the address clearing

File: checkout-ui-custom/src/partials/Deliver/Addresses.js [hunk 4]
```diff
@@ -24,12 +30,19 @@ const Addresses = () => {
...
     });
+};
 
-  return `
- <div class="bash--addresses shimmer" id="bash-address-list">
-    Loading address book...
-  </div>
-  `;
+const Addresses = () => {
+  RefreshAddressList();
```
Comment (shaheen-karodia) [inline checkout-ui-custom/src/partials/Deliver/Addresses.js:36]: There is a high chance that using the index DB methods to get Address's will mean we dont have to do this

File: checkout-ui-custom/src/partials/Deliver/EditAddressForm.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      </svg>
+      <p style="margin-left: 16px; line-height: 1; font-size: 14px; margin-bottom:0;">${addressLine}</p>
+  </div>
+  `;
+};
+
+const EditAddressForm = (data) => {
+  // VALIDATE DATA AND DEAL WITH ERRORS
+  // TODO HOW DO WE DETERMINE WHICH FORM TO SHOW
+  const fields = [
```
Comment (shaheen-karodia) [inline checkout-ui-custom/src/partials/Deliver/EditAddressForm.js:44]: could we get greater reusability of fields and reduce bundle size here by making fields generalizable

File: checkout-ui-custom/src/utils/services.js [hunk 6]
```diff
@@ -71,7 +71,7 @@ export const getAddresses = async () => {
 };
 
 // GET Address by ID / Name?
-const getAddress = async (addressName, fields) => {
+export const getAddress = async (addressName, fields) => {
```
Comment (shaheen-karodia) [inline checkout-ui-custom/src/utils/services.js:74]: we probably dont want to go down the master data level (and go straight to the index DB level)

**Review by giretyu — APPROVED** [review]
works fine my side 👌

**Review by simmbiote — APPROVED** [review]
lgtm 💪

**Review by SegopotsoMakhutja — APPROVED** [review]
Good on my side.

**Pull request merged**